src/common/image.c

```diff
@@ -2699,68 +2699,79 @@ int dt_image_local_copy_reset(const int32_t imgid)
 // xmp stuff
 // *******************************************************
 
-int dt_image_write_sidecar_file(const int32_t imgid)
+gboolean dt_image_write_sidecar_file(const int32_t imgid)
 {
-  // TODO: compute hash and don't write if not needed!
-  // write .xmp file
-  if((imgid > 0) && (dt_image_get_xmp_mode() != DT_WRITE_XMP_NEVER))
-  {
-    char filename[PATH_MAX] = { 0 };
+  if(imgid <= 0)
+    return TRUE;
 
-    // FIRST: check if the original file is present
-    gboolean from_cache = FALSE;
-    dt_image_full_path(imgid, filename, sizeof(filename), &from_cache);
+  const dt_imageio_write_xmp_t xmp_mode = dt_image_get_xmp_mode();
+  if(xmp_mode == DT_WRITE_XMP_NEVER)
+    return TRUE;
 
-    if(!g_file_test(filename, G_FILE_TEST_EXISTS))
-    {
-      // OTHERWISE: check if the local copy exists
-      from_cache = TRUE;
-      dt_image_full_path(imgid, filename, sizeof(filename), &from_cache);
+  if((xmp_mode == DT_WRITE_XMP_LAZY) && !dt_image_altered(imgid))
+    return TRUE;
 
-      //  nothing to do, the original is not accessible and there is no local copy
-      if(!from_cache) return 1;
-    }
+  char filename[PATH_MAX] = { 0 };
 
-    dt_image_path_append_version(imgid, filename, sizeof(filename));
-    g_strlcat(filename, ".xmp", sizeof(filename));
+  // FIRST: check if the original file is present
+  gboolean from_cache = FALSE;
+  dt_image_full_path(imgid, filename, sizeof(filename), &from_cache);
 
-    if(!dt_exif_xmp_write(imgid, filename))
-    {
-      // put the timestamp into db. this can't be done in exif.cc
-      // since that code gets called for the copy exporter, too
-      sqlite3_stmt *stmt;
-      DT_DEBUG_SQLITE3_PREPARE_V2
-        (dt_database_get(darktable.db),
-         "UPDATE main.images SET write_timestamp = STRFTIME('%s', 'now') WHERE id = ?1",
-         -1, &stmt, NULL);
-      DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
-      sqlite3_step(stmt);
-      sqlite3_finalize(stmt);
-      return 0;
-    }
+  if(!g_file_test(filename, G_FILE_TEST_EXISTS))
+  {
+    // OTHERWISE: check if the local copy exists
+    from_cache = TRUE;
+    dt_image_full_path(imgid, filename, sizeof(filename), &from_cache);
+
+    //  nothing to do, the original is not accessible and there is no local copy
+    if(!from_cache)
+      return TRUE;
+  }
+
+  dt_image_path_append_version(imgid, filename, sizeof(filename));
+  g_strlcat(filename, ".xmp", sizeof(filename));
+
+  if(!dt_exif_xmp_write(imgid, filename))
```

Comment:
This is introducing a regression, to reproduce:

- Ensure that you have "write sidecar file for each image" set to "after edit" & "look for updated XMP files on startup" activated.
- Take 2 images without XMP
- Import them
- Select one and go to darkroom, do an edit (change exposure)
- Quit darktable
- Restart it

And see that the DB timestamp has not been updated with the XMP timestamp.
![image](https://user-images.githubusercontent.com/467069/226189591-a22af4bf-e477-41fb-9ba8-53c1cfaaff76.png)

Reply:
I'll open a separate issue.